### PR TITLE
add `conan.tools.android.android_abi()`

### LINKED ISSUE #12814: [feature] Port conans.tools.to_android_abi to conan v2 tools
### What is your suggestion?

This function is used in opencv 4.x recipe of conan-center. It's the only legacy function remaining in https://github.com/conan-io/conan-center-index/pull/14947

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

### PR DESCRIPTION
**Pull request by SpaceIm**
Changelog: Feature: Add `conan.tools.android.android_abi()` function to return the Android standard ABI name based on Conan.
Docs: https://github.com/conan-io/docs/pull/2975

closes https://github.com/conan-io/conan/issues/12814

- [x] Refer to the issue that supports this Pull Request.
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [ ] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one. 

<sup>**Note:** By default this PR will skip the slower tests and will use a limited set of python versions. Check [here](https://github.com/conan-io/conan/blob/develop/.github/PR_INCREASE_TESTING.md) how to increase the testing level by writing some tags in the current PR body text.</sup>

### CODE REVIEW

File: conan/tools/android/utils.py
Comment (SSE4) [inline conan/tools/android/utils.py:30]: we're not handling x86 and MIPS variants here. at least x86 is popular for running Simulator builds (plus, there are real x86 boards with Android). MIPS was used by previous (old) NDK releases, but I would add for completeness.
(also I've heard Android is adding RISC-V now, but I don't know ABIs right now)

Comment (SpaceIm) [reply conan/tools/android/utils.py:30]: arch in profile and ndk abi are identical for x86 and mips.

Comment (SpaceIm) [reply conan/tools/android/utils.py:30]: it should be tested in unit test

Comment (SSE4) [reply conan/tools/android/utils.py:30]: oh, good catch! sorry for the noise :)

Comment (SpaceIm) [reply conan/tools/android/utils.py:30]: No it's fine, I've extended archs tested in the unit test.

**Review by SSE4 — APPROVED** [review]

**Comment by AbrilRBS** [comment]
I think this would be cool to have added in the docs :)

**Comment by SpaceIm** [comment]
Sure, but when I know it has a chance to be merged. I don't want to work on documentation if this feature is rejected.

File: conan/tools/android/utils.py
Comment (memsharded) [inline conan/tools/android/utils.py]: If we are providing the context, then falling back to the host settings doesn't look great, it should be explicit.
I ask you for the "build" context, you return the build context or fail. If this is being used in a cross-build scenario with 1 profile only, then it is allowed (should) fail.

**Review by memsharded — COMMENTED** [review]
I think this is looking good and could be merged, but I'd prefer to avoid a silent fallback to another context, better raise if profile build or target not correctly defined.

File: conans/test/unittests/tools/android/test_android_tools.py
Comment (AbrilRBS) [inline conans/test/unittests/tools/android/test_android_tools.py:10]: This is usually better to be done with a `@parametrized` decorator in the function itself, so failures are easier to debug, but I'm ok with this either way :)

File: conan/tools/android/utils.py
Comment (AbrilRBS) [reply conan/tools/android/utils.py]: I agree with @memsharded, this is something that should not fall back if missing. What usecase did you have in mind for allowing the fallback?

File: conan/tools/android/utils.py
Comment (AbrilRBS) [inline conan/tools/android/utils.py:31]: In the old code, `android_abi` would have been None for missing archs. The second parameter here changes that bahaviour, which might be more correct (No idea), but we should make sure it does not have any unintended side-effects

**Review by AbrilRBS — COMMENTED** [review]
Thanks for taking the time to implement this. I only have a few minor comments before we can merge :)

File: conans/test/unittests/tools/android/test_android_tools.py
Comment (SpaceIm) [reply conans/test/unittests/tools/android/test_android_tools.py:10]: I remember to have written tests with @pytest.mark.parametrize, then move to this implementation but I don't remember why. I can change back to parametrize.

File: conan/tools/android/utils.py
Comment (SpaceIm) [reply conan/tools/android/utils.py:31]: Well there are 2 old codes, inconsistent:
- legacy `toos.to_android_abi()` => it fallbacks to `str(conanfile.settings.arch)`
- CMakeToolchain => fallbacks to None

So a decision has to be made for this new one, as you wish, I've not a strong opinion on this.

File: conan/tools/android/utils.py
Comment (SpaceIm) [reply conan/tools/android/utils.py]: Yes I had 1 profile in mind, to not raise even if it makes little sense to call this function with 1 profile. Also I didn't want to be too strict with settings_target since it depends on context in which a recipe is called, I'm not sure of the consequences in recipes.

So what would you want? Raise `ConanException` if `settings_{context}` is not an attribute of conanfile? Even for settings_host?

File: conans/test/unittests/tools/android/test_android_tools.py
Comment (memsharded) [reply conans/test/unittests/tools/android/test_android_tools.py:10]: No big deal, if it works with parametrize easily, then do it, but don't waste time if you hit some issue.

File: conan/tools/android/utils.py
Comment (memsharded) [reply conan/tools/android/utils.py:31]: I think falling back to the arch input instead of ``None`` might be better, for cases where the users extend ``settings.yml`` with their own architectures, if they map directly to android ones, then this will work. So I am fine with this as-is.

File: conan/tools/android/utils.py
Comment (memsharded) [reply conan/tools/android/utils.py]: ``settings_host`` obviously doesn't exist, so it would be necessary to do the right thing, and do an ``if context == "host"`` or a dict mapping or something like that. But yes, it is better to raise if not defined, and this function should only be used with 2 profiles. It is typically better to be initially restrictive and raise, then letting some behavior slip, and then users depending on this behavior (for 1 profile) will complain later if this breaks for some reason that we need to force exclusively for 2 profiles (already happened a few times)

**Review by AbrilRBS — APPROVED** [review]
Thanks! (Sorry about the delay!)
Please don't forget about adding docs for this, will merge once linked so it's easier to write the changelog!

**Comment by memsharded** [comment]
We are releasing 1.58 asap, this would be missing ``settings`` raising when not defined, moving to 1.59

**Review by memsharded — COMMENTED** [review]
Apparently this is not that needed in ConanCenter, but I am fine with moving this forward for 1.59, if the settings fallback to host is removed.

**Review by memsharded — APPROVED** [review]

**Comment by SpaceIm** [comment]
Thanks for the final polishing. It will be useful in android-ndk recipe (but not anymore in opencv recipe, I have a PR removing its usage).

**Pull request merged**
